backend/cmm_helpers.ml

```diff
@@ -320,58 +320,337 @@ let add_no_overflow n x c dbg =
+               [arch_bits] times or more is the same as shifting [arch_bits - 1]
+               times *)
+            asr_const inner (Int.min (n + n') (arch_bits - 1)) dbg
+          | Cop (Clsr, [_; Cconst_int (n', _)], _)
+            when n' > 0 && is_defined_shift n' ->
+            (* If the argument is guaranteed non-negative, then we know the sign
+               bit is 0 and we can weaken this operation to a logical shift *)
+            lsr_const c1 n dbg
+          | Cop (Clsl, [c; Cconst_int (x, _)], _)
+            when is_defined_shift x && max_signed_bit_length c + x < arch_bits
+            ->
+            (* some operations always return small enough integers that it is
+               safe and correct to combine [asr (lsl x y) z] into [asr x (z -
+               y)]. *)
+            if x > n then lsl_const c (x - n) dbg else asr_const c (n - x) dbg
+          | Cop (Cor, [x; ((Cconst_int _ | Cconst_natint _) as y)], _) ->
+            or_int (asr_int x c2 dbg) (asr_int y c2 dbg) dbg
+          | Cop (Cand, [x; ((Cconst_int _ | Cconst_natint _) as y)], _) ->
+            and_int (asr_int x c2 dbg) (asr_int y c2 dbg) dbg
+          | Cop (Cxor, [x; ((Cconst_int _ | Cconst_natint _) as y)], _) ->
+            xor_int (asr_int x c2 dbg) (asr_int y c2 dbg) dbg
+          | _ -> Cop (Casr, [c1; c2], dbg)))
+      | Cop (Casr, [x; (Cconst_int (n', _) as y)], z), c2
+        when is_defined_shift n' ->
+        (* prefer putting the constant shift on the outside to help enable
+           further peephole optimizations *)
+        Cop (Casr, [Cop (Casr, [x; c2], dbg); y], z)
+      | _ -> Cop (Casr, [c1; c2], dbg))
+
+and lsl_int c1 c2 dbg =
+  map_tail2 c1 c2 ~f:(fun c1 c2 ->
+      match c1, c2 with
+      | c1, Cconst_int (0, _) -> c1
+      | c1, Cconst_int (n, _) when is_defined_shift n -> (
+        match get_const c1 with
+        | Some c1 -> natint_const_untagged dbg (Nativeint.shift_left c1 n)
+        | None -> (
+          match c1 with
+          | Cop (Clsl, [inner; Cconst_int (n', _)], dbg)
+            when is_defined_shift n' ->
+            if is_defined_shift (n + n')
+            then lsl_const inner (n + n') dbg
+            else replace inner ~with_:(Cconst_int (0, dbg))
+          | Cop (Caddi, [c1; Cconst_int (offset, _)], _)
+            when Misc.no_overflow_lsl offset n ->
+            add_const (lsl_int c1 c2 dbg) (offset lsl n) dbg
+          | Cop (Cor, [x; ((Cconst_int _ | Cconst_natint _) as y)], _) ->
+            or_int (lsl_int x c2 dbg) (lsl_int y c2 dbg) dbg
+          | Cop (Cand, [x; ((Cconst_int _ | Cconst_natint _) as y)], _) ->
+            and_int (lsl_int x c2 dbg) (lsl_int y c2 dbg) dbg
+          | Cop (Cxor, [x; ((Cconst_int _ | Cconst_natint _) as y)], _) ->
+            xor_int (lsl_int x c2 dbg) (lsl_int y c2 dbg) dbg
+          | c1 -> Cop (Clsl, [c1; c2], dbg)))
+      | Cop (Clsl, [x; (Cconst_int (n', _) as y)], z), c2
```

Comment:
```suggestion
      | Cop (Clsl, [x; (Cconst_int (n', _) as y)], dbg'), c2
```

this code is already tricky as is without having to remember that `z` is a debuginfo, contrary to `x` and `y`, ^^